File: HackerRank/Algorithms/Dynamic Programming/The_maximum_subarray.java
import java.util.Scanner;

public class Solution {

    public static int findMaximum(int a, int b) {
        if (a > b) {
            return a;
        }
        return b;
    }

    public static int findMaximum(int a, int b, int c) {
        return findMaximum(findMaximum(a, b), c);
    }

    public static void maximumSubArraySum(int[] a) {
        int maxContiguous    = a[0],
            maxNonContiguous = a[0],
            maxTillHere      = a[0];
        for(int i = 1; i < a.length; ++i) {
            maxTillHere      = findMaximum(maxTillHere + a[i], a[i]);
            maxContiguous    = findMaximum(maxContiguous, maxTillHere);
            maxNonContiguous = findMaximum(maxNonContiguous, a[i], maxNonContiguous + a[i]);
        }
        System.out.println(maxContiguous + " " + maxNonContiguous);
    }

    public static void main(String[] args) {
        /* Enter your code here.
           Read input from STDIN.
           Print output to STDOUT.
           Your class should be named Solution. */
        Scanner in = new Scanner(System.in);
        int testCase;
        testCase = in.nextInt();
        int size;
        int[] arr;
        while (testCase != 0) {
            size = in.nextInt();
            arr = new int[size];
            for(int i = 0; i < size; ++i) {
                arr[i] = in.nextInt();
            }
            maximumSubArraySum(arr);
            --testCase;
        }
        in.close();

    }
}

File: HackerRank/Algorithms/Dynamic Programming/The_maximum_subarray.py
#Python 3

def find_maximum_sub_array(a):
    max_contiguous = a[0]
    max_non_contiguous = a[0]
    max_till_here = a[0]
    for i in range(1, len(a)):
        max_till_here      = max(a[i], max_till_here + a[i])
        max_contiguous     = max(max_till_here, max_contiguous)
        max_non_contiguous = max(max(max_non_contiguous, a[i]), max_non_contiguous + a[i])
    print (max_contiguous, max_non_contiguous)

if __name__ == '__main__':
    test_case = int(input())
    while test_case:
        size = int(input())
        arr = [int(x) for x in input().split()]
        find_maximum_sub_array(arr)
        test_case -= 1
